Add Method to Convert QgsRasterBlock to NumPy Array

merydian opened the pull request:
This pull request, as part of qgis/QGIS-Enhancement-Proposals#227, introduces a new method to PyQGIS that enables conversion of QgsRasterBlock into a NumPy array. This addition allows users to efficiently handle and analyze raster data using NumPy, including the handling of no-data values through masked arrays. This enhancement improves integration with NumPy while keeping it an optional dependency. Closes #46324.

<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

merydian committed:
Implement QgsRasterBlock.as_numpy

Implement QgsRasterBlock.as_numpy

Small cleanup

Create test

Add type hints

Create qgis_data_type_to_numeric_data_type

Cleanup

nicogodet commented:
I think there is a missing entry in [CMakeList.txt](https://github.com/qgis/QGIS/blob/master/tests/src/python/CMakeLists.txt)

rouault commented on python/core/__init__.py.in:
```diff
@@ -43,6 +43,9 @@ from .additions.runtimeprofiler import ScopedRuntimeProfileContextManager
 from .additions.validitycheck import check
 from .additions.ranges import datetime_range_repr, date_range_repr
 
+import numpy
```
I'm not sure if numpy is a required dependency of QGIS python packages. Probably not. In that case it might be prudent to be robust to it missing

merydian replied:
Done. As per suggestion in the QEP.

rouault commented on python/core/__init__.py.in:
```diff
@@ -447,3 +450,35 @@ QgsRasterTransparency.TransparentThreeValuePixel.percentTransparent = property(g
+    Qgis.DataType.Int32: numpy.int32,
+    Qgis.DataType.Float32: numpy.float32,
+    Qgis.DataType.Float64: numpy.float64,
+    Qgis.DataType.CInt16: numpy.complex64,  # NumPy does not have complex int16, using complex64 as closest alternative
+    Qgis.DataType.CInt32: numpy.complex64,  # NumPy does not have complex int32, using complex64 as closest alternative
```
that will likely result in incorrect / garbage values. Better mapping to None and erroring out

nyalldawson replied:
:+1:  I'd go for raising a ValueError "The raster block datatype is not compatible with numpy arrays"

rouault commented on python/core/__init__.py.in:
```diff
@@ -447,3 +450,35 @@ QgsRasterTransparency.TransparentThreeValuePixel.percentTransparent = property(g
+    Qgis.DataType.CInt32: numpy.complex64,  # NumPy does not have complex int32, using complex64 as closest alternative
+    Qgis.DataType.CFloat32: numpy.complex64,
+    Qgis.DataType.CFloat64: numpy.complex128,
+    Qgis.DataType.ARGB32: numpy.uint32,  # NumPy does not have specific ARGB, using uint32 as closest alternative
```
an alternative would be to return a 3D numpy array with the first dimension being A,R,G or B

merydian replied:
I've set this to None for now. As with the CInt16/32.

rouault commented on python/core/__init__.py.in:
```diff
@@ -447,3 +450,35 @@ QgsRasterTransparency.TransparentThreeValuePixel.percentTransparent = property(g
+
+def raster_block_as_numpy(self) -> Union[numpy.ndarray, numpy.ma.MaskedArray]:
+   raster_dtype = qgis_data_type_to_numeric_data_type(self.dataType())
+   src_array = numpy.frombuffer(self.data(), dtype=raster_dtype)
```
.reshape() should be called to set the correct 2D shape

rouault commented on python/core/__init__.py.in:
```diff
@@ -447,3 +450,35 @@ QgsRasterTransparency.TransparentThreeValuePixel.percentTransparent = property(g
+    }
+   return qgis_to_numpy_dtype_dict[dataType]
+
+def raster_block_as_numpy(self) -> Union[numpy.ndarray, numpy.ma.MaskedArray]:
```
I would make the use of masked array something that could be disabled:
```suggestion
def raster_block_as_numpy(self, use_masking=True) -> Union[numpy.ndarray, numpy.ma.MaskedArray]:
```

nyalldawson commented on tests/src/python/test_qgsrasterblock.py:
```diff
@@ -0,0 +1,58 @@
+start_app()
+
+
+class TestQgsRasterBlock(QgisTestCase):
```
This test needs to be added to the Cmake file too (https://github.com/qgis/QGIS/blob/master/tests/src/python/CMakeLists.txt#L264), otherwise it won't be run

merydian committed:
Apply suggestions from code review

Small cleanup

Set numpy equivalent of ARGB32 to None

Add test for masking=False

Make masking optional

Reshape array to be 2D

Adapt test to 2D array

Adapt test line in docstring

Add test to CMakeLists.txt

Wrap import and method setting in try/except

Raise ValueError when dataType is not supported

Set numpy equivalent for CInt16 and CInt32 to None

Apply str method to dataType in Error message

rouault commented on python/core/__init__.py.in:
```diff
@@ -447,3 +449,43 @@ QgsRasterTransparency.TransparentThreeValuePixel.percentTransparent = property(g
+       Qgis.DataType.CFloat32: numpy.complex64,
+       Qgis.DataType.CFloat64: numpy.complex128,
+       Qgis.DataType.ARGB32: None,
+       Qgis.DataType.ARGB32_Premultiplied: numpy.uint32  # Same as ARGB32
```
```suggestion
       Qgis.DataType.ARGB32_Premultiplied: None
```

merydian replied:
Right, thanks!

merydian committed:
Set numpy equivalent for ARGB32_Premultiplied to None

Co-authored-by: Even Rouault <even.rouault@spatialys.com>

nyalldawson commented on python/core/__init__.py.in:
```diff
@@ -447,3 +449,43 @@ QgsRasterTransparency.TransparentThreeValuePixel.percentTransparent = property(g
+   QgsRasterBlock.as_numpy = raster_block_as_numpy
+
+except ModuleNotFoundError:
+   pass
```
I suggest we raise an exception when numpy isn't available, but only when the function is called. (This mimics what we do in other PyQGIS geometry methods when the GEOS library version is too old for a called function). 

E.g.:


```suggestion
   def raster_block_as_numpy(self, use_masking=True):
       raise QgsNotSupportedException('QgsRasterBlock.as_numpy is not available, numpy is not installed on the system')
       
    QgsRasterBlock.as_numpy = raster_block_as_numpy
```

nyalldawson commented on python/core/__init__.py.in:
```diff
@@ -447,3 +449,43 @@ QgsRasterTransparency.TransparentThreeValuePixel.percentTransparent = property(g
+         raise ValueError(f"The raster block data type '{str(self.dataType())}' is not compatible with NumPy arrays.")
+      src_array = numpy.frombuffer(self.data(), dtype=raster_dtype)
+      src_array = src_array.reshape((self.height(), self.width()))
+      noDataValue = self.noDataValue() if isinstance(self.noDataValue(), raster_dtype) else 0
+      if not self.hasNoDataValue() or not use_masking:
+         return src_array
+      else:
+         return numpy.ma.masked_equal(src_array, noDataValue)
```
```suggestion      
      if not self.hasNoDataValue() or not use_masking:
         return src_array
      else:
         no_data_value = self.noDataValue() if isinstance(self.noDataValue(), raster_dtype) else 0
         return numpy.ma.masked_equal(src_array, no_data_value)
```

nyalldawson commented on tests/src/python/test_qgsrasterblock.py:
```diff
@@ -0,0 +1,62 @@
+
+class TestQgsRasterBlock(QgisTestCase):
+    def setUp(self):
+        self.iface = get_iface()
```
We don't need iface for these tests:

```suggestion
```

nyalldawson commented:
It's **extremely** gross, but right now you'll have to copy and paste this change into the Qt 6 bindings (https://github.com/qgis/QGIS/tree/master/python/PyQt6/core) too :upside_down_face:  

That's why the Qt6 test is failing here...

merydian committed:
Copy feature to Qt6 bindings

merydian committed:
Raise exception when numpy isn't available and function is called

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

merydian committed:
Move setting of nodata Value to else statement

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

merydian committed:
Remove iface from test setUp

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

merydian committed:
Add as_numpy not available exception to Qt6 bindings

merydian committed:
Correct indentation

merydian committed:
Adapt no_data_value setting in Qt6 bindings

merydian committed:
Add import of type hints

nyalldawson commented on tests/src/python/test_qgsrasterblock.py:
```diff
@@ -0,0 +1,61 @@
+)
+import unittest
+from qgis.testing import start_app, QgisTestCase
+from qgis.testing.mocked import get_iface
```
```suggestion
```

nyalldawson committed:
Update tests/src/python/test_qgsrasterblock.py

nyalldawson commented:
@rouault are you happy with this version now?

nyalldawson commented:
Gentle ping @rouault
